lister.cc

```diff
@@ -75,3 +77,44 @@ future<> lister::rmdir(fs::path dir) {
         return remove_file(dir.native());
     });
 }
+
+future<std::optional<directory_entry>> queue_lister::get() {
+    if (!_done) {
+        auto f = co_await open_checked_directory(general_disk_error_handler, _dir.native());
+        auto walker = [this] (fs::path dir, directory_entry de) {
+            return _q.push_eventually(std::make_optional<directory_entry>(std::move(de))).handle_exception([] (std::exception_ptr) {
+                // ignore exception, queue was aborted
```

Comment:
If I understand correctly, this means that if the queue is aborted, the lister will continue to list potentially a million more files, and ignore them one by one. Wouldn't it be better to pass the exception through, and presumably the lister will stop listing?

Reply:
It might be possible, but requires adding support for properly aborting the lister.

Reply:
> It might be possible, but requires adding support for properly aborting the lister.

I was assuming (without checking the code) that if the walker throws an exception, the lister will properly abort. Doesn't it?
Your code deliberately catches the exception that the walker throws. I was thinking, maybe if it did throw - everything will work as expected?

Reply:
it did throw and all hell broke loose :)
I'm exaggerating a bit, but I wish it worked as simple as returning an exception from the walker.

Reply:
So maybe you discovered a bug in the existing lister, which should be fixed?

The user-written "walker" is supposed to return a future, and this future **may** be exceptional, and the lister needs to support this case in a reasonable way - and if it doesn't, I would think it's a bug. 

NOTE: I'm not suggesting that walker should throw (it is probably called without futurize() protection), but rather that it can return an exceptional future. I think this is what happens in coroutines automatically.

Reply:
Fixed in v2